doc/crossedit.man
.\"$Id$
.TH xledit 0.89.3 "Mar 7 1993"
.SH NAME
xledit - A map-editor for the game Crossfire.
.SH SYNOPSIS
.B xledit
[-debug] [-pix] [-e <map_level> | <height>x<width>]
.SH DESCRIPTION
.PP
.I xtedit
enables you to design maps for the game Crossfire.
.SH OPTIONS
.TP
.B -debug
Enable debug
.TP
.B -pix
Use pixmaps instead of fonts.
.TP
.TP
.B -e
Specify the level of the map to edit after this argument.
Alternatively, if it is a new level, you can specify its size
in the format "<height>x<width>"
.SH DESIGNING MAPS
If you start crossfire with the -e <map-level> option, you will enter
the editor and be allowed to change a map.  If the level you specified
doesn't exist, you will start creating a new (remember to specify the
name of the level before clicking on the save button).
.PP
To change a map, select items from the list of archetypes by moving the
scrollbar with the left and right mouse-button, and then clicking on any
item.  You can also click on any of the backgrounds displayed above
the list of archetypes.  After you have selected something, you can
put it into the map by pressing the left mousebutton.  To put something
on top of another thing, use the right mouse button.
To remove all items on one spot, first choose the background which is
empty (the last one) and then click the left button somewhere on
the map.  If you use the right instead, only the top item will be
deleted.
You can do any of these things repeatedly by holding down the button in
question and move the mouse around.
To scroll the map-window, use the cursor keys or press the middle button
on the mouse while pointing on the scrollbars.
.PP
To save a map, click on the save button.  Check that you have specified
the map-name before doing this (in case you started editing on a new map).
To load another map, just click in the "Level:" button and type in a new
level (followed by return) and then click on the load button.
The width and height of a map can be changed by clicking on the respective
button and typing in a new number (followed by return).  The position that
the players enters the map on can be changed by clicking on either of the
"start" buttons and then clicking someplace on the map.
.PP
To change where an exit-object leads, click on it (after having placed it
on the map) with the middle mouse button.  Then change "Level:", "X:" and "Y:"
by clicking and typing.  If it's level is 0, it automatically leads to
the next level.  If X and Y is 0, the players will start in the default
start positions on the next level.
Some other items can also be edited in the same fashion, like gates, buttons,
handles, holes, teleporters, etc.
.SH FILES
.TP
.B LIBDIR/archetypes
Global objects which can be put on the maps are stored here.
.TP
.B LIBDIR/mapindex
A list of all maps, and their corresponding number is kept here.
.TP
.B LIBDIR/maps/*.om
The files which contains the background of the maps.
.TP
.B LIBDIR/maps/*.oo
The files which contains the objects of the maps.
.TP
.SH SEE ALSO
crossfire
.SH BUGS
Please let me know about any bugs you find in the game.  I know there are
plenty of them, and some are mentioned on the TODO file which is supplied
with the game.
.SH AUTHOR
Copyright (C) 1992 Frank Tore Johansen (frankj@ifi.uio.no)
.PP
Authors and contributors to this program:
.LP
.RS
.ft B
.nf
frankj@ifi.uio.no (Frank Tore Johansen)
kjetilho@ifi.uio.no (Kjetil Torgrim Homme)
tvangod@ecst.csuchico.edu (Tyler)
elmroth@cd.chalmers.se (Tony Elmroth)
dougal.scott@fcit.monasu.edu.au (Dougal Scott)
wchuang@athena.mit.edu (William)
ftww@cs.su.oz.au (Geoff Bailey)
jorgens@flipper.pvv.unit.no (Kjetil Wiekhorst J\(/orgensen)
c.blackwood@rdt.monash.edu.au (Cameron Blackwood)
jtraub+@cmu.edu (Joseph L. Traub)
rgg@aaii.oz.au (Rupert G. Goldie)
eanders+@cmu.edu (Eric A. Anderson)
eneq@Prag.DoCS.UU.SE (Rickard Eneqvist)
Jarkko.Sonninen@lut.fi (Jarkko Sonninen)
kholland@sunlab.cit.cornell.du (Karl Holland)
vick@bern.docs.uu.se (Mikael Lundgren)
mol@meryl.csd.uu.se (Mikael Olsson)
.fi
.ft R
.RE
.LP
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.

.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.

.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

.\" The author can be reached via e-mail to frankj@ifi.uio.no.


CHANGES
Changes in top level CVS tree.  A few notes when putting entries in this
file:
1) Please include a date in the form mm/dd/yyyy (yes, us americans are 
screwed up relative to most the world, but some consistent date format
is needed for this file.)
2) You don't need to go into exacting detail of the changes.  This file
is mostly meant to give an idea to others what has changed - you don't
need to recreate the code here.  If you change fixes a reported bug
on the mailing list, please provide a brief description of the bug
fixed.  But please do try to include a useful description of
what changed.  Saying something like 'General cleanup - just whitespace
changes' is much more useful than 'misc edits' or 'code cleanup'.
3) Please include your name with the changes you make.
4) Whenever making an update, please update this file, no matter how
small - this just lets others know that the file has changed if nothing
else.  With this, include the file(s) that you changed.
------------------------------------------------------------------------------

Changes for CVS:

This checkin mainly fixes multipart teleporters - they should now
work properly.  Also a minor fix to make the new pickup code work.
server/apply.c: apply_shop_mat rewritten to be more 'presentable'.
    Minor changes were needed in any case to pass the player/creature
    a a second object to teleport.
server/c_object.c: change the sscanf to be unsigned when getting value
    for new pickup mode.  At least on my system, if the value in the
    string was larger than the max signed value, sscanf set it to the 
    max value.
server/move.c: teleport function largely re-written.  Instead of the
    third arg being the 'originator', it is now the object to teleport.
    This allows the head of the teleporter to get passed to this function
    even if the object to teleport is not on the head.
server/time.c: move_teleporter modified to also check for objects on the
    other parts of multipart teleporters.  Also, pass the head object
    to the teleport function so that proper destination is used.
MSW 2001-09-02

Add Nils Lohner patch that new selections to the pickup mode.

include/define.h: Pickup type defines added.
include/player.h: Change mode (pickup mode) to 32 bits.
server/c_object.c: command_pickup function modified to support all
    the new pickup options.
server/player.c: Fix for path_to_player to prevent crashes.  Add check
    in move_player_attack just in case.  check_pick function modified to
    support new pickup options.  do_some_living modified to eat flesh
    items if player is starving and they don't have any food.
server/spell_effect.c: cast_detection modified so that detect magic/curse
    will not show objects beneath the floor.  Change so that detect_monster
    and detect_evil spells copy the face of what they detected - gives better
    appearance on client.
socket/item.c: fix bug LookAt that ws checking blocked_los on wrong
    offset.
MSW 2001-08-28

Enhancements to make monsters 'do the right thing' with relation
to tiled maps (ie, will detect players and attack them if on an adjoining
tiled maps).  This involved a rewrite of path_to_player - I've tested this
out, and if anything, it seems that monsters may be more aggressive on
spellcasting.
++
common/map.c: blocked modify to make sure it is checking valid coordinate
   value.  blocked_link,blocked_two modified to translate for tiled maps.
   Add load_and_link_tiled_map which loads a tiled map and links in all
   pointers relative to adjoining maps.  out_of_map modified to use
   this function instead of doing the work itself.  Add get_rangevector
   function that returns offset, direction, and distance between two
   objects, taking into account tiled maps.  get_rangevector_from_mapcoord
   added which is similar to get_rangevector, but takes a map and coordinates
   instead of a second object.  on_same_map function added which returns
   true if the two objects are on the same logical (ie tiled) maps.
doc/crossfire.doc: Add note about what Int does for monsters.
include/map.h: Add some comments, and add structure for rangevector
   function return values.
server/attack.c: add calls to on_same_map instead of doing op->map
   comparisons.
server/monster.c: Many functions modified to properly handle tiled
   maps properly.
server/pets.c: get_pet_enemy modified to know about tiled maps.
server/player.c: get_nearest_player modified to know about tiled maps.
   path_to_player rewritten to know about tiled maps - I think some other 
   bugs were also fixed in return values.  Split player_attack_door
   from move_player attack to keep functions a more reasonable size.
   move_player_attack, player_can_view modified to know about tiled
   maps.
server/skills.c: can detect_enemy modified to take another parameter -
   call in steal updated to pass this parameter.
server/spell_util.c: ok_to_put_more, can_see_monsterP modified to
   handle tiled maps properly.
socket/item.c: look_at modified to return proper results when player
   looks at something on an adjoining tiled map.
++ End of checkin

common/object.c:decrease_ob_nr modified to correctly update client when
    player uses item from container that is on the ground (bookshelf for
    example)
server/time.c: move_teleporter modified to skip over wall/floors and
    teleport whatever is above them.
MSW 2001-08-05

common/map.c: modify link_multipart_objects to preserve name and title of
    head when making the more parts.  modify update_position so that it should
    never show an invisible face to the player.
common/object.c: fix check_walk_on so that spells will now properly get
    affected by spinners and directories.
include/sockproto.h: Rebuilt to include MapNewmapCmd
server/hiscore.c: Fix crash if player savebeds map on map which does not
    have a name.  Use the path instead.
server/move.c: modify teleporter function to send MapNewmapCmd to client

crossedit/crossedit.c
/*
	Copyright (C) 1993 Jarkko Sonninen & Petri Heinila

	This program is free software; you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation; either version 2 of the License, or
	(at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program; if not, write to the Free Software
	Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
	
	The authors can be reached via e-mail to Jarkko.Sonninen@lut.fi
	or Petri.Heinila@lut.fi .
*/

#include <Posix.h>
#include "App.h"
#include "X11.h"
#include "debug.h"

/**********************************************************************
 * application resources 
 **********************************************************************/

#define Offset(field) XtOffsetOf(AppRes, field)
static XtResource resources[] = {
    {"mapWidth", "MapWidth", XtRDimension, sizeof (Dimension),
    Offset (mapWidth), XtRDimension, NULL},
    {"mapHeight", "MapHeight", XtRDimension, sizeof (Dimension),
    Offset (mapHeight), XtRDimension, NULL},
    {"usePixmaps", "UsePixmaps", XtRBoolean, sizeof (XtRBoolean),
    Offset (usePixmaps), XtRImmediate, False},
    {"useColorPixmaps", "UseColorPixmaps", XtRBoolean, sizeof (XtRBoolean),
     Offset (useColorPixmaps), XtRImmediate, False},
    {"usePng", "UsePng", XtRBoolean, sizeof (XtRBoolean),
     Offset (usePng), XtRImmediate, False},
    {"cmdCrossfire", "CmdCrossfire", XtRString, sizeof (XtRString),
    Offset (cmdCrossfire), XtRString, NULL},
    {"cmdReset", "CmdReset", XtRString, sizeof (XtRString),
    Offset (cmdReset), XtRString, NULL},
    {"creator", "Creator", XtRString, sizeof (XtRString),
    Offset (creator), XtRString, "No Body"},
    {"email", "Email", XtRString, sizeof (XtRString),
    Offset (email), XtRString, "none@foo.bar"},
};
#undef Offset

static XrmOptionDescRec options[] = {
    {"-pix","*usePixmaps",XrmoptionNoArg,"TRUE"},
    {"-xpm","*useColorPixmaps",XrmoptionNoArg,"TRUE"},
    {"-png","*usePng",XrmoptionNoArg,"TRUE"},

}; 

String fallback_resources[] = {
#include <Crossedit.ad.h>
    NULL
};

enum DisplayMode displaymode=Dm_Png;

void EditDirtyAc(Widget w, XEvent * event, String * argv, Cardinal * argc) {
    debug0 ("EditDirtyAc\n");
}

void Nop(Widget w, XEvent * event, String * argv, Cardinal * argc) {}
XtActionsRec mainActions[] = {
  {"Nop", Nop},
  {"EditKey", EditDirtyAc},
};

/**********************************************************************
 * main
 **********************************************************************/

int main (int argc, char **argv, char **env)
{
    XtAppContext app_con;

    App app;

/* Crossedit seems to have so many memory problems, this seems like
 * a really good idea.
 */
#ifdef DEBUG_MALLOC_LEVEL
  malloc_debug(DEBUG_MALLOC_LEVEL);
#endif


    /* crossfire init */
    init_globals();
    init_library ();
    init_regions();
    init_archetypes ();
    editor = TRUE;
#if 0
    /* Enable this if you want people to be able to edit maps
     * in a different directory other than the master map
     * directory.  Note that the 'editor' directory of maps
     * needs to be in LIBDIR/maps (LIBDIR/maps being where
     * the maps will be saved.
     *
     * An alternative method would be to have a link to a high
     * level directory from the master map source (ie, something
     * like 'ln -s /users .LIBDIR.  Then users can select the users
     * directory in the map editor, which will take them to the
     * directory above their home directory, select their home directory,
     * and then whatever directory they want to save the maps in.
     */
    if (getenv("LIBDIR")) LibDir = getenv("LIBDIR");
#endif

    /*XtAppInitialize (&app_con, "Crossedit", NULL, 0,  &argc, argv, 
		     fallback_resources, NULL, 0);*/

    XtToolkitInitialize ();
    app_con = XtCreateApplicationContext();    
#if 1
    XtAppSetFallbackResources (app_con, fallback_resources);
#endif
    XtAppAddActions(app_con,mainActions, XtNumber (mainActions));

    /*XtGetApplicationResources (app.shell, (XtPointer) & app->res,
			       resources, XtNumber (resources), NULL, 0);*/

    app = AppCreate(app_con, "Crossedit", resources, XtNumber (resources),
		    options,XtNumber(options), &argc, argv);


    /*** usage ***/
    if (argc > 1) {
      fprintf(stderr,"Usage: crossedit [-options]\n\t[-options]:\tXt-options\n\n");
      exit(2);
    }

    XtRealizeWidget (app->shell);
    XtAppMainLoop (app_con);
    exit(0);
}

/*** End of crossfire.c ***/
